File: clean-code/chapters/chapter-10-classes/listing_10_7.py
class TablePrinter:

    def __init__(self, rows_per_page, columns_per_page, page_header):
        self.__rows_per_page = rows_per_page;
        self.__columns_per_page = columns_per_page;
        self.__page_header = page_header;
        self.__numbers_per_page = rows_per_page * columns_per_page;
    
    def print(self, data):
        page_number = 1
        first_index_on_page = 0
        while (first_index_on_page < len(data)):
            last_index_on_page = min(first_index_on_page + self.__numbers_per_page - 1, len(data) - 1)
            self.__print_page_header(page_number)
            self.__print_page(first_index_on_page, last_index_on_page, data)
            print("")
            page_number += 1
            first_index_on_page += self.__numbers_per_page


    def __print_page_header(self, page_number):
        print(f"{self.__page_header} --- Page {page_number}")
        print("")

    def __print_page(self, first_index_on_page, last_index_on_page, data):
        first_index_of_last_row_on_page = first_index_on_page + self.__rows_per_page - 1
        for first_index_in_row in range(first_index_on_page, first_index_of_last_row_on_page + 1):
            self.__print_row(first_index_in_row, last_index_on_page, data)
            print("")

    def __print_row(self, first_index_in_row, last_index_on_page, data):
        for column in range(self.__columns_per_page):
            index = first_index_in_row + (column * self.__rows_per_page)
            if (index <= last_index_on_page):
                print(f"{data[index]}", end = " ")

File: clean-code/chapters/chapter-10-classes/listing_10_7.java
package literatePrimes;

import java.io.PrintStream;

public class RowColumPagePrinter {
    private int rowsPerPage;
    private int columnsPerPage;
    private int numbersPerPage;
    private String pageHeader;
    private PrintStream printStream;

    public RowColumnPagePrinter(int rowsPerPage,
                                int columnsPerPage,
                                String pageHeader) {
        this.rowsPerPage = rowsPerPage;
        this.columnsPerPage = columnsPerPage;
        this.pageHeader = pageHeader;
        this.numbersPerPage = rowsPerPage * columnsPerPage;
        this.printStream = System.out;
    }

    public void print(int data[]) {
        int pageNumber = 1;
        for (int firstIndexOnPage = 0;
             firstIndexOnPage < data.length;
             firstIndexOnPage += this.numbersPerPage) {
            int lastIndexOnPage = 
                Math.min(firstIndexOnPage + numbersPerPage - 1, data.length - 1);
            printPageHeader(this.pageHeader, pageNumber);
            printPage(firstIndexOnPage, lastIndexOnPage, data);
            this.printStream.println("\f");
            pageNumber++;      
        }   
    }

    private void printPage(int firstIndexOnPage,
                           int lastIndexOnPage,
                           int[] data) {
        int firstIndexOfLastRowPage = 
            firstIndexOnPage + this.rowsPerPage - 1;
        for (int firstIndexInRow = firstIndexOnPage;
             firstIndexInRow <= firstIndexOfLastRowPage;
             firstIndexInRow++) {
            printRow(firstIndexInRow, lastIndexOnPage, data);
            this.printStream.println("");
        }
    }

    private void printRow(int firstIndexInRow,
                          int lastIndexOnPage,
                          int[] data) {
        for (int column = 0; column < this.columnsPerPage; column++) {
            int index = firstIndexInRow + column * this.rowsPerPage;
            if (index <= lastIndexOnPage) {
                this.printStream.format("%10d", data[index]);
            }
        }
    }

    private void printPageHeader(String pageHeader,
                                 int pageNumber) {
        this.printStream.println(pageHeader + " --- Page " + pageNumber);
        this.printStream.println("");
    }

    public void setOutput(PrintStream printStream) {
        this.printStream = printStream;
    }
}